### ISSUE TITLE
Error installing SAIGE SyntaxError: Missing parentheses in call to 'print'

### ISSUE THREAD

**Issue by hhx037** [weizhouUMICH/SAIGE#160]
Hi,

I'm running into an issue when attempting to install SAIGE in R 3.6.1 using devtools (other methods fail too), and switching from Python 3 to Python 2 does not help, though the error is different:

Python 3.6.3

```
Successfully installed xiaoyeli/superlu
Waf: The wscript in '/tmp/RtmpBbyak3/R.INSTALLb24133ca2226/SAIGE/thirdParty/bgen' is unreadable
Traceback (most recent call last):
  File "/tmp/RtmpBbyak3/R.INSTALLb24133ca2226/SAIGE/thirdParty/bgen/.waf3-2.0.6-f8aee10f7730da942389d88501d489e8/waflib/Scripting.py", line 101, in waf_entry_point
    set_main_module(os.path.normpath(os.path.join(Context.run_dir,Context.WSCRIPT_FILE)))
  File "/tmp/RtmpBbyak3/R.INSTALLb24133ca2226/SAIGE/thirdParty/bgen/.waf3-2.0.6-f8aee10f7730da942389d88501d489e8/waflib/Scripting.py", line 141, in set_main_module
    Context.g_module=Context.load_module(file_path)
  File "/tmp/RtmpBbyak3/R.INSTALLb24133ca2226/SAIGE/thirdParty/bgen/.waf3-2.0.6-f8aee10f7730da942389d88501d489e8/waflib/Context.py", line 360, in load_module
    exec(compile(code,path,'exec'),module.__dict__)
  File "/tmp/RtmpBbyak3/R.INSTALLb24133ca2226/SAIGE/thirdParty/bgen/wscript", line 33
    print "Creating %s build..." % bld.options.mode
                               ^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print("Creating %s build..." % bld.options.mode)?
Waf: The wscript in '/tmp/RtmpBbyak3/R.INSTALLb24133ca2226/SAIGE/thirdParty/bgen' is unreadable
Traceback (most recent call last):
  File "/tmp/RtmpBbyak3/R.INSTALLb24133ca2226/SAIGE/thirdParty/bgen/.waf3-2.0.6-f8aee10f7730da942389d88501d489e8/waflib/Scripting.py", line 101, in waf_entry_point
    set_main_module(os.path.normpath(os.path.join(Context.run_dir,Context.WSCRIPT_FILE)))
  File "/tmp/RtmpBbyak3/R.INSTALLb24133ca2226/SAIGE/thirdParty/bgen/.waf3-2.0.6-f8aee10f7730da942389d88501d489e8/waflib/Scripting.py", line 141, in set_main_module
    Context.g_module=Context.load_module(file_path)
  File "/tmp/RtmpBbyak3/R.INSTALLb24133ca2226/SAIGE/thirdParty/bgen/.waf3-2.0.6-f8aee10f7730da942389d88501d489e8/waflib/Context.py", line 360, in load_module
    exec(compile(code,path,'exec'),module.__dict__)
  File "/tmp/RtmpBbyak3/R.INSTALLb24133ca2226/SAIGE/thirdParty/bgen/wscript", line 33
    print "Creating %s build..." % bld.options.mode
                               ^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print("Creating %s build..." % bld.options.mode)?
ERROR: configuration failed for package ‘SAIGE’
* removing ‘/data/home/hhx037/R/x86_64-pc-linux-gnu-library/3.6/SAIGE’
Error: Failed to install 'SAIGE' from GitHub:
  (converted from warning) installation of package ‘/tmp/RtmpbBijWr/file4a131691efa9/SAIGE_0.36.3.2.tar.gz’ had non-zero exit status
```

Python 2.7.14

```
*** installing help indices
** building package indices
** testing if installed package can be loaded from temporary location
Error: package or namespace load failed for ‘SAIGE’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/data/home/hhx037/R/x86_64-pc-linux-gnu-library/3.6/00LOCK-SAIGE/00new/SAIGE/libs/SAIGE.so':
  /data/home/hhx037/R/x86_64-pc-linux-gnu-library/3.6/00LOCK-SAIGE/00new/SAIGE/libs/SAIGE.so: undefined symbol: sgecon_
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/data/home/hhx037/R/x86_64-pc-linux-gnu-library/3.6/SAIGE’
Error: Failed to install 'SAIGE' from GitHub:
  (converted from warning) installation of package ‘/tmp/Rtmp0OovaL/file8c0f4fb3e9aa/SAIGE_0.36.3.2.tar.gz’ had non-zero exit status
```

Is it because the print statements in some of your scripts are still in Python2 format?

Kind regards,

Stephane

**Comment by weizhouUMICH** [comment]
Hi Stephane, 

I think the first error with python 3 is from installing the bgen library as the main SAIGE script is written in R and C++ not python. Not sure why the python 2 gives the error. 
It would be much easier to try use the conda environment file with the devtools to install SAIGE (approach 5b https://github.com/weizhouUMICH/SAIGE)

Thanks!
Wei

**Comment by hhx037** [comment]
Hi Wei,

thanks for your prompt reply. I've just tried with a conda environment (anaconda2/5.2.0), and it doesn't work for me either (both with 5c devtools and 5b, same error):

```
installing to /data/home/hhx037/R/x86_64-pc-linux-gnu-library/3.6/00LOCK-SAIGE/00new/SAIGE/libs
** R
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded from temporary location
Error: package or namespace load failed for ‘SAIGE’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/data/home/hhx037/R/x86_64-pc-linux-gnu-library/3.6/00LOCK-SAIGE/00new/SAIGE/libs/SAIGE.so':
  /data/home/hhx037/R/x86_64-pc-linux-gnu-library/3.6/00LOCK-SAIGE/00new/SAIGE/libs/SAIGE.so: undefined symbol: sgecon_
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/data/home/hhx037/R/x86_64-pc-linux-gnu-library/3.6/SAIGE’
```

Any idea?

Cheers,

Steph

**Comment by weizhouUMICH** [comment]
Hi Steph, 
I've never seen this error before. Can you try delete the SAIGE.o and *.so files and re-compile?

Thanks!
Wei

**Comment by matuskosut** [comment]
Hi @weizhouUMICH, what is the state of `bgen` code included in 3rdparty? do you sync it from other repo e.g. https://enkre.net/cgi-bin/code/bgen/dir?ci=trunk ? or do you manage it separately as a fork? 

I noticed that in the original repo for example `bgen/wscript` seems to be already ported to python 3 (https://enkre.net/cgi-bin/code/bgen/file?name=wscript&ci=trunk). Would it make sense to use that work and push this code to 2020? Especially since python 2.7 is not supported anymore..

**Comment by weizhouUMICH** [comment]
We have just released a new version 1.0.0. It has computational efficiency improvements for both Step 1 and Step 2 for single-variant and set-based tests. We have created a new program github page https://github.com/saigegit/SAIGE with the documentation provided https://saigegit.github.io/SAIGE-doc/
The program will be maintained by multiple SAIGE developers there. The docker image has been updated. Please feel free to try the version 1.0.0 and report issues if any.

Thanks!
Wei
